src/reducers/language.js

```diff
@@ -0,0 +1,79 @@
+import { CHANGE_CONST, CHANGE_FUN, CHANGE_PRED } from '../actions';
+import {parseConstants, parseFunctions, parsePredicates} from '@fmfi-uk-1-ain-412/js-fol-parser';
+
+const language = (state = { const: { input: "", object: new Set(), error: ""}, fun: { input: "", object: new Map(), error: "" }, pred: { input: "", object: new Map(), error: "" }, factories: {} }, action = {type: undefined}) => {
+    switch (action.type) {
+        case CHANGE_CONST:
+            let newState = {
+                const: {
+                    input: action.text,
+                    object: state.const.object,
+                    error: state.const.error
+                },
+                fun: state.fun,
+                pred: state.pred
+            }
+            try {
+                let res = undefined;
+                if (action.text !== "") {
+                    res = parseConstants(action.text); 
+                }
+                newState.const.object = new Set(res);
+                newState.const.error = "";
+            } catch (e) {
+                newState.const.error = e;
+            }
+            return newState;

```

Comment:
Parser by mal zvladnut aj prazdny zoznam konstant.

Z hladiska stability kodu v buducnosti je lepsie co najviac pouzivat spread syntax a dotykat sa iba tych casti stavu, ktorych sa naozaj musite. V tomto pripade to znamena, ze nebudete objekt `const` konstruovat so vsetkymi atributmi, ale spravite spreadom kopiu a zmenite len to, co treba menit. A uz vobec by ste v casti, kde sa zaoberate konstantami, nemali spominat `fun` a `pred`.

Mimochodom `const` je nestastny nazov, lebo je to klucove slovo JavaScriptu. Objekt sice moze mat takyto atribut, ale nemozete mat premennu/konstantu s tymto menom. Lepsie by bolo `consts` (a `funs`, `preds`) – nakoniec mnozne cislo (-s) je na mieste.

Kod sa potom da zostrucnit takto:
```javascript
            let consts = {...state.consts, input: action.text};
            try {
                const res = parseConstants(action.text);
                consts.object = new Set(res);
                consts.error = "";
            } catch (e) {
                consts.error = e;
            }
            return {...state, consts};
```
V poslednom returne je vyuzite, ze `{consts}` je skratka za `{consts: consts}`.

Keby v buducnosti do celeho stavu jazyka alebo do stavu `consts` cosi pribudlo, takto upraveneho kodu sa to vobec nedotkne.

Reply:
Po vymazani inputu parser vrati hlasku "SyntaxError: Expected constant identifier but end of input found."

Reply:
Aha, pardon. Tak nechajte ten test na mieste. Konstanty nie su povinne, ani funkcie. Jedine pri predikatoch je OK, ak sa objavi chyba, ked nie su ziadne – lebo nejaky v jazyku musi byt.